RFC.md

```diff
@@ -0,0 +1,207 @@
+# [RFC] MOE design in Torchtune
+
+## Background
+This RFC proposes adding the MOE support in Torchtune. We want to design in a general way so that components can be easily swapped when implementing different MOE models. An MOE layer directly replaces the dense FFN layer in the transformer decoder layer and has two main components: router and experts.
+
+## Expert
+An expert is essentially an FFN layer similar to the original dense FFN layer in the transformer decoder layer. There are two kinds of experts: routed experts and shared experts. Each expert in the routed experts specializes in learning certain patterns/aspects, and only part of the routed experts will be activated. On the other hand, shared experts are always activated, aiming at capturing and consolidating common knowledge across varying contexts.
+
+**Here's the proposed Experts design in torchtune:**
+```python
+class Experts(nn.Module):
```

Comment:
Naming, I think this should be something like MoELinear

Reply:
+1 here

Reply:
curious what's the reason? I was thinking having Experts, Routers, and MoELayers is more clear in terms of what each class is doing. Otherwise, there's no Experts class and this `MoELinear` is essentially Experts? I don't have any strong preferences, but just want to know why you think MoELinear is better?

Reply:
It matches our own convention of defining FeedForward (singular) rather than a group of things like Experts.

Reply:
But here we are defining a group of experts/feedforward. Wouldn't `MoELinear` be confusing? It's more like `GroupedMoELinear`?